quadtree/quadtree.go

```diff
@@ -509,6 +546,48 @@ func (v *nearestVisitor) Visit(p geo.Pointer) {
 	}
 }
 
+type nearestVisitorGeo struct {
+	point        *geo.Point
+	filter       Filter
+	k            int
+	closest      pointsQueue
+	closestBound *geo.Bound
+	distance     float64
+}
+
+func (v *nearestVisitorGeo) Bound() *geo.Bound {
+	return v.closestBound
+}
+
+func (v *nearestVisitorGeo) Point() *geo.Point {
+	return v.point
+}
+
+func (v *nearestVisitorGeo) Visit(p geo.Pointer) {
+	// skip this pointer if we have a filter and it doesn't match
+	if v.filter != nil && !v.filter(p) {
+		return
+	}
+
+	point := p.Point()
+	if d := point.GeoDistanceFrom(v.point); d < v.distance {
+		heap.Push(&v.closest, pointsQueueItem{point: p, distance: d})
+		if v.closest.Len() > v.k {
+			heap.Pop(&v.closest)
+
+			// Actually this is a hack. We know how heap works and obtain top element without function call
+			top := v.closest[0]
+
+			v.distance = top.distance
+
+			// We have filled queue, so we start to restrict searching range
+			x := v.point.X()
+			y := v.point.Y()
+			v.closestBound.Set(x-v.distance, x+v.distance, y-v.distance, y+v.distance)
```

Comment:
So I don't think this is going to work. The bound needs to have have the correct min/max lat/lon given the distance. Here v.distance (meters) is being subtracted from x (lat/lon) which is a units mismatch and will lead to wrong results.